### ISSUE TITLE
[core] Add a matrix option

### ISSUE THREAD

**Issue by aalmiray** [jreleaser/jreleaser#1774]
**Is your feature request related to a problem? Please describe.**
Languages with cross-compile (such as Go, Rust) support are capable of producing binaries for multiple chips on the same platform. The `ArchiveAssembler` is bound to the current platform. It would be better if its filename and inputs were parameterizable with a "Github Actions" inspired matrix.

**Describe the solution you'd like**
```
assemble:
  archive:
    helloworld:
      active: ALWAYS
      formats: [ ZIP ]
      matrix:
        - { goos: darwin,  arch: arm64, platform: osx-aarch_64     }
        - { goos: darwin,  arch: amd64, platform: osx-x86_64       }
        - { goos: linux,   arch: arm64, platform: linux-aarch_64   }
        - { goos: linux,   arch: amd64, platform: linux-x86_64     }
        - { goos: windows, arch: arm64, platform: windows-aarch_64 }
        - { goos: windows, arch: amd64, platform: windows-x86_64   }
      attachPlatform: false
      archiveName: '{{distributionName}}-{{projectVersion}}-{{ matrix.goos }}-{{ matrix.arch }}'
      fileSets:
        - input: 'target/{{ matrix.goos }}-{{ matrix.arch }}'
          output: 'bin'
          includes: [ 'helloworld{.exe,}' ]
        - input: '.'
          includes: [ 'LICENSE' ]
```

**Describe alternatives you've considered**
Invoking the `assemble` command multiple times in different nodes.

**Comment by aalmiray** [comment]
The `matrix.platform` key has a special meaning and will define the resolved artifact's platform when filling up information on the related distribution.

**Comment by aalmiray** [comment]
This feature has been upgraded to a full section. Hooks can also benefit from matrix definitions. Here's a full example on how a matrix may be used to build 6 binaries, assemble 6 archives, and configure a single distribution, with cross-compiled Go code

```yaml
matrix:
  rows:
    - { goos: darwin,  goarch: arm64, platform: osx-aarch_64     }
    - { goos: darwin,  goarch: amd64, platform: osx-x86_64       }
    - { goos: linux,   goarch: arm64, platform: linux-aarch_64   }
    - { goos: linux,   goarch: amd64, platform: linux-x86_64     }
    - { goos: windows, goarch: arm64, platform: windows-aarch_64 }
    - { goos: windows, goarch: amd64, platform: windows-x86_64   }

hooks:
  script:
    before:
      - run: |
          go build -o target/{{ matrix.goos }}-{{ matrix.goarch }}/ src/helloworld.go
        applyDefaultMatrix: true
        environment:
          GOOS: '{{ matrix.goos }}'
          GOARCH: '{{ matrix.goarch }}'
        filter:
          includes: ['assemble']
          
project:
  name: helloworld
  description: HelloWorld in Go
  longDescription: HelloWorld in Go
  links:
    homepage: https://github.com/jreleaser/helloworld-go
  authors:
    - Andres Almiray
  license: MIT
  inceptionYear: 2024
  stereotype: CLI
    
release:
  github:
    name: helloworld-go
    overwrite: true
    changelog:
      formatted: ALWAYS
      preset: conventional-commits
      contributors:
        format: '- {{contributorName}}{{#contributorUsernameAsLink}} ({{.}}){{/contributorUsernameAsLink}}'
        
assemble:
  archive:
    helloworld:
      active: ALWAYS
      formats: [ ZIP ]
      applyDefaultMatrix: true
      archiveName: '{{distributionName}}-{{projectVersion}}-{{ matrix.goos }}-{{ matrix.goarch }}'
      fileSets:
        - input: 'target/{{ matrix.goos }}-{{ matrix.goarch }}'
          output: 'bin'
          includes: [ 'helloworld{.exe,}' ]
        - input: '.'
          includes: [ 'LICENSE' ]

distributions:
  helloworld:
    type: BINARY
    executable:
      windowsExtension: exe
```

**Comment by aalmiray** [comment]
🎉 This issue has been resolved in `v1.16.0` ([Release Notes](https://github.com/jreleaser/jreleaser/releases/tag/v1.16.0))
